### ISSUE TITLE
Out of memory issue with many sequences

### ISSUE THREAD

**Issue by mashu** [mashu/LineageCollapse.jl#14]
Stack trace is
```julia
OutOfMemoryError()                                                                                                                                                               
Stacktrace:                                                                                                                                                                     
  [1] GenericMemory                                                                                                                                                             
    @ ./boot.jl:516 [inlined]                                                                                                                                                   
  [2] new_as_memoryref                                                                                                                                                          
    @ ./boot.jl:535 [inlined]                                                                                                                                                   
  [3] Array                                                                                                                                                                     
    @ ./boot.jl:582 [inlined]                                                                                                                                                   
  [4] Array                                                                                                                                                                     
    @ ./boot.jl:592 [inlined]                                                                                                                                                   
  [5] zeros                                                                                                                                                                     
    @ ./array.jl:578 [inlined]                                                                                                                                                  
  [6] zeros                                                                                                                                                                     
    @ ./array.jl:574 [inlined]                                                                                                                                                  
  [7] compute_pairwise_distance(metric::NormalizedHammingDistance, sequences::Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}) 
  ```
  This happened after some time with one of bigger libraries, presumably hitting large cluster.
  
  The idea is to use sparse array with zeros in lower triangle and Float32, code below works just fine
  ```julia
  using SparseArrays
using LinearAlgebra
using Clustering
dist = spzeros(Float32, 10, 10)
for i in 1:10
    for j in i+1:1
        dist[i, j] = 1
    end
end
sdist = LinearAlgebra.Symmetric(dist)
hclusters = hclust(sdist, linkage=:single)
```
The `Symmetric` wrapper from LinearAlgebra is to return upper values when accessing bottom triangle.

**Comment by mashu** [comment]
Help needed to fix test in [fix_memory_usage](https://github.com/mashu/LineageCollapse.jl/tree/fix_memory_usage) @mchernys

**Comment by mashu** [comment]
Ok, this works but obviously sparse matrix is not memory safe. So perhaps some other solution is needed.

**Comment by mashu** [comment]
Ok this is fixed by simply performing hierarchical clustering only for uniqu CDR3s and than mapping back to original sequences. Since same CDR3s end-up in the same cluster anyway, there is no need for redundant calculations.

Fixed by 9899b7cc5f710f1c5c16d4b5cb1f13fa38f5eae5
